File: androidApp/src/main/java/org/multipaz/wallet/android/ui/document/FloatingItemListClaim.kt
package org.multipaz.wallet.android.ui.document

import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.CancellationException
import kotlinx.datetime.TimeZone
import kotlinx.serialization.json.jsonPrimitive
import org.multipaz.claim.Claim
import org.multipaz.claim.JsonClaim
import org.multipaz.claim.MdocClaim
import org.multipaz.compose.decodeImage
import org.multipaz.compose.items.FloatingItemHeadingAndContent
import org.multipaz.compose.items.FloatingItemHeadingAndText
import org.multipaz.documenttype.DocumentAttributeType
import org.multipaz.util.fromBase64Url

/**
 * A composable item that displays a claim within a floating item list.
 *
 * If the claim is of type [DocumentAttributeType.Picture], it decodes the image bytes and renders
 * the image. Otherwise, it renders the claim value as text.
 *
 * @param claim The [Claim] to display.
 * @param heading The heading to display above the claim value, defaulting to [Claim.displayName].
 * @param modifier The modifier to apply to the item.
 * @param timeZone The time zone to use when rendering date/time values, defaulting to current system default.
 */
@Composable
fun FloatingItemListClaim(
    claim: Claim,
    heading: String = claim.displayName,
    modifier: Modifier = Modifier,
    timeZone: TimeZone = TimeZone.currentSystemDefault(),
) {
    if (claim.attribute?.type == DocumentAttributeType.Picture) {
        val img = try {
            val bytes = when (claim) {
                is MdocClaim -> claim.value.asBstr
                is JsonClaim -> claim.value.jsonPrimitive.content.fromBase64Url()
            }
            decodeImage(bytes)
        } catch (err: Exception) {
            if (err is CancellationException) throw err
            FloatingItemHeadingAndText(
                heading = heading,
                text = "Image decoding: $err",
                modifier = modifier
            )
            null
        }
        if (img != null) {
            FloatingItemHeadingAndContent(
                heading = heading,
                modifier = modifier,
                content = {
                    Image(
                        bitmap = img,
                        contentDescription = null,
                        alignment = Alignment.TopStart,
                        modifier = Modifier.fillMaxWidth().size(250.dp),
                    )
                }
            )
        }
    } else {
        val str = claim.render(timeZone)
        FloatingItemHeadingAndText(
            heading = heading,
            text = str,
            modifier = modifier
        )
    }
}

File: iosApp/iosApp/FloatingItemListClaim.swift
import SwiftUI
import Multipaz

struct FloatingItemListClaim: View {
    let claim: Claim
    var heading: String? = nil
    var timeZone: Multipaz.TimeZone = Multipaz.TimeZone.Companion.shared.currentSystemDefault()

    private var resolvedHeading: String {
        heading ?? claim.displayName
    }

    var body: some View {
        if claim.attribute?.type is DocumentAttributeType.Picture {
            if let uiImage = decodeImage(claim: claim) {
                FloatingItemHeadingAndContent(
                    heading: resolvedHeading,
                    content: {
                        Image(uiImage: uiImage)
                            .resizable()
                            .aspectRatio(contentMode: .fit)
                            .frame(maxWidth: 250, maxHeight: 250)
                    }
                )
            } else {
                FloatingItemHeadingAndText(
                    heading: resolvedHeading,
                    text: "Error decoding image"
                )
            }
        } else {
            FloatingItemHeadingAndText(
                heading: resolvedHeading,
                text: claim.render(timeZone: timeZone)
            )
        }
    }

    private func decodeImage(claim: Claim) -> UIImage? {
        var data: Data? = nil
        if let mdocClaim = claim as? MdocClaim {
            var value = mdocClaim.value
            if let tagged = value as? Tagged {
                value = tagged.taggedItem
            }
            if let bstr = value as? Bstr {
                data = bstr.asBstr.toNSData() as Data
            }
        } else if let jsonClaim = claim as? JsonClaim {
            if let content = jsonClaim.value.jsonPrimitive.content as? String {
                data = decodeBase64Url(content)
            }
        }

        if let data = data {
            return UIImage(data: data)
        }
        return nil
    }

    private func decodeBase64Url(_ string: String) -> Data? {
        var base64 = string
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")

        let mod = base64.count % 4
        if mod > 0 {
            base64 += String(repeating: "=", count: 4 - mod)
        }

        return Data(base64Encoded: base64)
    }
}
